src/cloud_providers/aws/pricing.rs

```diff
@@ -24,40 +28,333 @@ impl PricingClient {
+                    warn!("Attempt {} failed, will retry", retry_count);
+                }
+            }
+        }
+
+        error!("All retries failed. Last error: {:?}", last_error);
+        None
+    }
+
+    /// Single attempt to fetch EC2 pricing data
+    ///
+    /// # Arguments
+    /// * `filters` - Vector of filters to apply to the pricing query
+    ///
+    /// # Returns
+    /// * `Result<Option<FlattenedData>, Box<dyn Error>>` - Result containing either:
+    ///   - Ok(Some(data)) - Successfully found pricing data
+    ///   - Ok(None) - No matching instances found
+    ///   - Err(e) - An error occurred during the request
+    async fn attempt_get_ec2_price(
+        &self,
+        filters: Vec<PricingFilters>,
+    ) -> Result<Option<FlattenedData>, Box<dyn std::error::Error + Send + Sync>> {
+        // Create paginated request to AWS Pricing API
         let mut response = self
             .client
             .get_products()
-            .service_code("AmazonEC2".to_string())
-            .set_filters(Some(filters))
-            .into_paginator()
+            .service_code("AmazonEC2".to_string()) // Specifically query EC2 prices
+            .set_filters(Some(filters)) // Apply the filters (instance type, OS, etc)
+            .into_paginator() // Handle pagination of results
             .send();
 
         let mut data = Vec::new();
 
-        while let Some(Ok(output)) = response.next().await {
+        // Process each page of results
+        while let Some(output) = response.next().await {
+            let output = output?; // Propagate any AWS API errors
+
+            // Process each product in the current page
             for product in output.price_list() {
-                let pricing: PricingData = serde_json::from_str::<Query<PricingData>>(product)
-                    .unwrap()
-                    .into();
-                let flat_data = FlattenedData::flatten_data(&pricing);
-                data.push(flat_data);
+                // Parse the JSON pricing data using serde_query
+                match serde_json::from_str::<Query<PricingData>>(product) {
+                    Ok(pricing) => {
+                        // Convert the complex pricing data into a flattened format
+                        let flat_data = FlattenedData::flatten_data(&pricing.into());
+                        data.push(flat_data);
+                    }
+                    Err(e) => {
+                        error!("Failed to parse product data: {:?}", e);
+                        continue; // Skip invalid products
+                    }
+                }
             }
         }
 
-        data.into_iter().reduce(|a, b| {
-            if a.price_per_unit > b.price_per_unit {
-                a
-            } else {
-                b
-            }
-        })
+        debug!("Processed pricing data length: {}", data.len());
+
+        // Return the most expensive instance from the results
+        if data.is_empty() {
```

Comment:
This is still a draft but i wanted to point out that, `reduce` returns None on an empty iterator.

Reply:
Yes, i know but that would be more clear for new engineers that read the code and don't know rust very well like me. And we are not adding complexity.